File: Leetcode/Python3/1020. Number of Enclaves/solution.py
# https://leetcode.com/problems/number-of-enclaves/
from collections import deque
dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        def bfs(x, y):
            close, move = True, 1
            q = deque()
            q.append((x, y))
            grid[x][y] = 0
            while q:
                x, y = q.popleft()
                # If the land is connected to the wall of the grid, it is not close.
                if x == 0 or x == m - 1 or y == 0 or y == n - 1:    
                            close = False
                for i in range(4):
                    nx = x + dx[i]
                    ny = y + dy[i]
                    if 0 <= nx < m and 0 <= ny < n and grid[nx][ny] == 1:
                        q.append((nx, ny))
                        grid[nx][ny] = 0
                        move += 1
            return move if close else 0
        m, n = len(grid), len(grid[0])
        moves = 0
        for idx_x in range(m):
            for idx_y in range(n):
                if grid[idx_x][idx_y] == 1:
                    moves += bfs(idx_x, idx_y)
        return moves

File: Leetcode/C++/1020. Number of Enclaves/solution.cpp
// https://leetcode.com/problems/number-of-enclaves/description/

#include <vector>

using namespace std;

class Solution
{
public:
    int numEnclaves(vector<vector<int>> &grid)
    {
        vector<int> dx = { 0, 0, 1, -1 }, dy = { 1, -1, 0, 0 };
        int area = 0;

        for (int y = 0; y < grid.size(); y++)
        {
            for (int x = 0; x < grid[0].size(); x++)
            {
                if (grid[y][x] == 1)
                {
                    int count = 1;
                    bool enclose = true;
                    vector<pair<int, int>> stack = { { y, x } };
                    grid[y][x] = 0;
                    while (!stack.empty())
                    { // DFS
                        int cur_y = stack.back().first, cur_x = stack.back().second;
                        stack.pop_back();

                        for (int idx = 0; idx < dx.size(); idx++)
                        {
                            int ny = cur_y + dy[idx], nx = cur_x + dx[idx];
                            if (ny < 0 || ny >= grid.size() || nx < 0 || nx >= grid[0].size())
                            {
                                enclose = false;
                                continue;
                            }
                            if (grid[ny][nx] == 1)
                            {
                                stack.push_back({ ny, nx });
                                grid[ny][nx] = 0;
                                count += 1;
                            }
                        }
                    }
                    if (enclose)
                    { // Add to area if enclosed
                        area += count;
                    }
                }
            }
        }
        return area;
    }
};